src/runtime/pyobject.cs

```diff
@@ -182,17 +177,47 @@ public T As<T>()
         /// </remarks>
         protected virtual void Dispose(bool disposing)
         {
-            if (!disposed)
+            if (this.obj == IntPtr.Zero)
+            {
+                return;
+            }
+
+            if (Runtime.Py_IsInitialized() == 0)
+                throw new InvalidOperationException("Python runtime must be initialized");
+
+            if (!Runtime.IsFinalizing)
             {
-                if (Runtime.Py_IsInitialized() > 0 && !Runtime.IsFinalizing)
+#if DEBUG
+                long refcount = Runtime.Refcount(this.obj);
+                Debug.Assert(refcount > 0);
+
+                bool checkError = refcount == 1;
+                IntPtr errType = IntPtr.Zero, errVal = IntPtr.Zero, traceback = IntPtr.Zero;
+                if (checkError)
                 {
-                    IntPtr gs = PythonEngine.AcquireLock();
-                    Runtime.XDecref(obj);
-                    obj = IntPtr.Zero;
-                    PythonEngine.ReleaseLock(gs);
+                    Runtime.PyErr_Fetch(out errType, out errVal, out traceback);
                 }
-                disposed = true;
+#endif
+
+                Runtime.XDecref(this.obj);
+
+#if DEBUG
+                if (checkError)
+                {
+                    if (errType != IntPtr.Zero || errVal != IntPtr.Zero || traceback != IntPtr.Zero)
+                    {
+                        // Python requires finalizers to preserve exception:
+                        // https://docs.python.org/3/extending/newtypes.html#finalization-and-de-allocation
+                        Runtime.PyErr_Restore(errType, errVal, traceback);
+                    }
+                    else
+                    {
+                        Runtime.CheckExceptionOccurred();
```

Comment:
Not a good practice for using Fetch and Restore in this case.
It makes a delay raise if some exceptions occurred in `this.obj.ob_type.tp_dealloc` on here.
case 1: Dispose of object A raises an exception, then it pass to next object B till here to raise.
case 2: Dispose of object A raises an exception, but it's the last object in finalize queue, it will be raise on a unpredictable position.
We can assume(or expect) there's no exception before calling `Decref(obj)`, setting an assert check before starting collect may help for debug build.

Reply:
```c#
bool checkError =  Runtime.Refcount(this.obj) == 1;
Runtime.XDecref(this.obj);
if (checkError)
{
    Runtime.CheckExceptionOccurred();
}
```
Just enough for here.

Reply:
> Dispose of object A raises an exception, then it pass to next object B till here to raise.

But that can't happen in this code. Object's A error will be thrown in line 215 before object B's decref will even happen.

Just doing `Runtime.CheckExceptionOccurred()` is not enough due to the requirement to preserve set exception I mentioned before.

Reply:
No, it won't, if the error occurred in `Runtime.XDecref(this.obj);` and no errors before calling this.

Reply:
@amos402 can you re-read the code? Either I don't understand your comment, or you've read the code incorrectly.

Assuming object A has refcount 1, then `checkError == true` for object A:

- line 198 `Runtime.PyErr_Fetch` will fill all error variables with `IntPtr.Zero`, now no error is set
- line 202 will call decref on A
- line 207 `if` condition `== false`, because all error variable are `IntPtr.Zero` from line 198, so it will go to `else` branch
- line 215, if decref from line 202 raised an error, it will be thrown here immediately

That happens when there's no pending error upon entering `Dispose`. If there's one, line 207 condition will instead be `true`, the potential error from decrefing A will be ignored, and the original error restored instead.

Reply:
```c#
#if DEBUG
                long refcount = Runtime.Refcount(this.obj);
                Debug.Assert(refcount > 0);

                bool checkError = refcount == 1; // refcoucnt == 1
                IntPtr errType = IntPtr.Zero, errVal = IntPtr.Zero, traceback = IntPtr.Zero;
                if (checkError) // true
                {
                    Runtime.PyErr_Fetch(out errType, out errVal, out traceback);
                    // errType != IntPtr.Zero
                }
                disposed = true;
#endif
                Runtime.XDecref(this.obj);
                // Exception raised in Runtime.XDecref(this.obj), or add following code for simulate
                Runtime.PyErr_SetString(Exceptions.Exception, "dealloc exception");
#if DEBUG
                if (checkError) // true
                {
                    // errType != IntPtr.Zero
                    if (errType != IntPtr.Zero || errVal != IntPtr.Zero || traceback != IntPtr.Zero)
                    {
                        // Enter here, exception will be override and it will raise next time.

                        // Python requires finalizers to preserve exception:
                        // https://docs.python.org/3/extending/newtypes.html#finalization-and-de-allocation
                        Runtime.PyErr_Restore(errType, errVal, traceback);
                    }
                    else
                    {
                        Runtime.CheckExceptionOccurred();
                    }
                }
```

Reply:
@amos402 but `errType != IntPtr.Zero` can only be when some exception is already set. In which case simply doing `Runtime.CheckExceptionOccurred` will just raise that exception, which we should not do because "Python requires finalizers to preserve exception"                     https://docs.python.org/3/extending/newtypes.html#finalization-and-de-allocation

Reply:
Not for here, exception may set before the first Dispose. That's why I said `We can assume(or expect) there's no exception before calling Decref(obj), setting an assert check before starting collect may help for debug build.`
And if with that precondition, fetch and restore become meaningless.

Reply:
@amos402 it is unclear why can we assume there's no exception before calling `Decref`.

Reply:
@amos402 in fact, I can see we could face a problem with that assumption. Suppose you wants to mix high- and low-level wrappers, and make something like:

```csharp
using (dynamic uri = Py.Import("uri")){
  if (uri.fetch() == null) { Exceptions.SetError(NotFound, "address not found"); }
}
DoSomethingElse();
```
Without `PyErr_Fetch` in `Dispose` this code will raise `PythonException` when exiting `using` block, so `DoSomethingElse` will never be called.

Reply:
If we preserved exception and give it to Python till Python call some byte code or Python check the exception set or not, the exception may raise in the wrong place. My previous sample(https://gist.github.com/amos402/5eb5db264bf63414db395fdc4fd178e1) code shown that. Thus, we must handle these exceptions immediately, raise them or make a PyFatalError abort the program.

Why "Python requires finalizers to preserve exception" didn't match here?
Because we custom handle the object's lifecycle by `Runtime.PyObject`, yet not in Python interpreter in `ceval.c`, it will check the exception set on every byte code boundary(not 100% correct, because internal function in CPython may check and raise) and raise them to the user code.

When should we accept "Python requires finalizers to preserve exception"?
After we handled the exception.
You can see the example code from document:
```c
cbresult = PyObject_CallObject(self->my_callback, NULL);
        if (cbresult == NULL)
            PyErr_WriteUnraisable(self->my_callback);
```
it handled by print the exception information. Well, you can say it ignore the exception too, but at least it left some information not just ignore totally.
If 

Sample:
```c#
IntPtr result = Runtime.PyDict_GetItemWithError();
// error occurred, result == IntPtr.Zero, exception set;
// developer ignore the result and continue.
var xxx = new PyObejct(); // just exactly the timing finalizer call `Collect`,
// finish create `xxx`, the exception still remain, when it would revealed to user? Hard to say, maybe after this scope program return to the Python interpreter and Python raise it, or maybe after calling some C function it check the exception, like `assert(PyErr_Occurred());` or `if (PyErr_Occurred()) { // some stuff }`, no matter which kind, the true exception position is just far from the raise position, that's what we want to see.
```

Reply:
@amos402 I can add PyErr_WriteUnraisable, but I think my example shows why we have to do fetch and restore in `Dispose`

Reply:
> 
> 
> @amos402 in fact, I can see we could face a problem with that assumption. Suppose you want to mix high- and low-level wrappers, and make something like:
> 
> ```cs
> using (dynamic uri = Py.Import("uri")){
>   if (uri.fetch() == null) { Exceptions.SetError(NotFound, "address not found"); }
> }
> DoSomethingElse();
> ```
> 
> Without `PyErr_Fetch` and `Restore` in `Dispose` this code will raise `PythonException` when exiting `using` block, so `DoSomethingElse` will never be called.

Yes, that assumption should build on a program which is bug free for this part. So `setting an assert check before starting collect may help for debug build.` is as a requirement.

Reply:
@amos402 the sample that I posted has no bugs. In fact, I think because of that sample we should do fetch and restore in release mode too.

Reply:
> 
> 
> @amos402 I can add `PyErr_WriteUnraisable`, but I think my example shows why we have to do fetch and restore in `Dispose`

If the exception handled, fetch and restore is good.

> 
> 
> @amos402 the sample that I posted has no bugs. In fact, I think because of that sample we should do fetch and restore in release mode too.

Not really, no bugs if we handle the exception, and fetch & restore can guard the exception not be affected, but the raise position of r the exception may still far from it's exactly position. Plus every `Dispose` plus two C calls. Wouldn't it be simple to check exception and raise at start before we extract the object from queue? Then we can assume it's crystal clear before all `Dispose`. 😆I have say that I changed my mind for putting the check on debug build only, it should also check on release build.

Reply:
@amos402 instead of using `PyErr_WriteUnraisable` I just throw .NET exception, but preserve the Python one.

Reply:
> If user got an exception set somehow, why does it have to be `Finalizer` to reveal it? This is a completely unrelated problem, and should not be our concern in the first place. If really necessary, user should be able to call `PyErr_Fetch` or a higher level wrapper at any time they want.

It may not related, but it may help. You can search the usages of `assert(!PyErr_Occurred());` on CPython code, same principle. Well, I always trend to reduce the PInvoke calls because of their costs. So it's better if we can avoid to call fetch & restore.

If just keep the `PyErr_Fetch` and call `CheckExceptionOccurred`, when exception occurred in `Decref`, the exception will lost, it will not preserve them to Python. Exceptions left to the local variables of this frame.

Reply:
> So it's better if we can avoid to call fetch & restore.

We can't due to the example I provided:
```csharp
using (dynamic uri = Py.Import("uri")){
  if (uri.fetch() == null) { Exceptions.SetError(NotFound, "address not found"); }
}
DoSomethingElse();
```

> If just keep the `PyErr_Fetch` and call `CheckExceptionOccurred`, when exception occurred in `Decref`, the exception will lost, it will not preserve them to Python

Why are you saying that? The original Python exception is restored for Python in `finally` block. And the exception from `Decref` is raised as a C# exception, which the caller of `Dispose` can handle in whatever way they want.

Reply:
Aha, just too for this conversation, didn't notice that you wrap it by finally, now it should be fine.